src/Lumper.UI/Updater/Updater.cs

```diff
@@ -0,0 +1,406 @@
+                await downloadStream.CopyToAsync(stream, cts.Token);
+            }
+            else
+            {
+                int read;
+                var length = (int)response.Content.Headers.ContentLength.Value;
+                var remaining = length;
+                while (!handler.Cancelled &&
+                       (read = await downloadStream.ReadAsync(
+                           buffer.AsMemory(0, int.Min(buffer.Length, remaining)),
+                           cts.Token)) >
+                       0)
+                {
+                    var prog = (float)read / length * DownloadProgressPercentage;
+                    handler.UpdateProgress(prog, $"{float.Floor((1 - ((float)remaining / length)) * DownloadProgressPercentage)}%");
+                    await stream.WriteAsync(buffer.AsMemory(0, read), cts.Token);
+                    remaining -= read;
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            if (ex is TaskCanceledException)
+                Logger.Info("Download cancelled by user");
+            else
+                Logger.Error(ex, "Failed to download!");
+
+            await stream.DisposeAsync();
+            return null;
+        }
+        finally
+        {
+            stream.Close();
+        }
+
+        if (handler.Cancelled)
+            return null;
+
+        return stream;
+    }
+    /// <summary>
+    /// Returns the URL to the download link for the OS-specific version.
+    /// </summary>
+    /// <returns></returns>
+    private static string GetPath(GithubUpdate assets, OSPlatform OS)
+    {
+        var operatingSystem = "";
+
+        if (OS == OSPlatform.Windows)
+            operatingSystem = "win";
+        else if (OS == OSPlatform.Linux)
+            operatingSystem = "linux";
+        else
+            throw new ApplicationException("Unsupported OS: " + OS);
+
+        foreach (Asset asset in assets.Assets)
+        {
+            if (asset.Name.Contains(operatingSystem, StringComparison.OrdinalIgnoreCase))
+            {
+                return asset.BrowserDownloadUrl;
+            }
+
+        }
+        throw new Exception("Could not find download link for " + nameof(OS));
+    }
+
+    /// <summary>
+    /// Downloads a file from the given URI and places it in fileName
+    /// </summary>
+    public static async Task DownloadFile(Uri uri, string fileName, IoProgressWindow progressWindow, IoHandler handler, CancellationTokenSource cts) => await HttpDownload(uri.AbsoluteUri, fileName, progressWindow,  handler, cts);
+    /// <summary>
+    /// Downloads an update for the program, applies it, and then restarts itself with the new version
+    /// </summary>
+    /// <returns></returns>
+    public static async ValueTask Update()
+    {
+        GithubUpdate assets;
+        try
+        {
+            assets = await GetGithubUpdates();
+
+            Version? latest;
```

Comment:
Should be `Version latest = GetVersionFromString(assets.TagName);` note that latest is non-null. Idk if your editor has LSP supported but this should be warning at you about this, don't ignore it!